src/btree/bt_prefetch.c

```diff
@@ -95,11 +96,20 @@ __wt_prefetch_page_in(WT_SESSION_IMPL *session, WT_PREFETCH_QUEUE_ENTRY *pe)
 
     WT_STAT_CONN_INCR(session, block_prefetch_pages_read);
 
+    /*
+     * If pre-fetching for a checkpoint handle, ensure the page-in checkpoint reader visibility
```

Comment:
The code change looks fine, but I don't really understand this comment. Could you elaborate on what "worried" means in this context and "our visibility setup"?

Reply:
I believe in this context the threads which push the checkpoint-related pages onto the prefetch queue and pop the pages off the queue to read will usually be different threads. However, I think this means that they have different visibility setups (session->txn) and it is problematic if they do. Checkpoint expects the WT_TXN_IS_CHECKPOINT flag to be set .

Reply:
@korteland I'm proposing a comment like this, but holding off until we figure out the exact fix we want and do some testing. This draft assumes the approach that Hari is suggesting.

When reading from checkpoint files, all readers will use the same checkpoint snapshot where it is assumed that anything visible to the snapshot is considered globally visible. It is likely that the original thread which pushes checkpoint-related refs onto the pre-fetch queue will be a different thread to the one which eventually pops the ref/s off the queue and reads the corresponding pages into memory. If the transaction contexts of these two threads differs (e.g. the pre-fetch thread does not have the same WT_TXN_IS_CHECKPOINT state), the page read logic will fail as checkpoint-related pages go through a visible all check when they are read into memory. Skip this visible all check when we read checkpoint-related pages using a session which enables pre-fetching as it is okay to read these contents.

Reply:
Thanks for the followup @mm-ng - that's a great writeup. If we do go with Hari's approach, I'd LGTM that comment.

Reply:
I axed this comment in the newer commit [3a516b7](https://github.com/wiredtiger/wiredtiger/pull/9925/commits/3a516b74fa137c3c5dff4526234cc7c5cdd5764a) :-( but let me know if you feel like the new comment is lacking details.